cpp/unreal_plugins/SpEngine/Source/SpEngine/WorkQueue.h

```diff
@@ -0,0 +1,113 @@
+        io_context_.run();
+
+        // reinitialize io_context_ and executor_work_guard_ to prepare for the next call to run()
+        mutex_.lock();
+        io_context_.restart();
+        new(&executor_work_guard_) boost::asio::executor_work_guard<boost::asio::io_context::executor_type>(io_context_.get_executor());
+        mutex_.unlock();
+    }
+
+    void reset()
+    {
+        // request io_context_.run() to stop executing once all of its scheduled work is finished
+        mutex_.lock();
+        executor_work_guard_.reset();
+        mutex_.unlock();
+    }
+
+    template <typename TFunc>
+    static auto wrapFuncToExecuteInWorkQueueBlocking(WorkQueue& work_queue, TFunc&& func)
+    {
+        return wrapFuncToExecuteInWorkQueueBlockingImpl(work_queue, std::forward<TFunc>(func), FuncInfo<TFunc>());
+    }
+
+private:
+
+    template <typename TFunc, typename TReturn, typename... TArgs> requires CFuncReturnsAndIsCallableWithArgs<TFunc, TReturn, TArgs...>
+    static auto wrapFuncToExecuteInWorkQueueBlockingImpl(WorkQueue& work_queue, TFunc&& func, const FuncInfo<TReturn(*)(TArgs&...)>& fi)
+    {
+        // Note that we capture func by value because we want to guarantee that func is still
+        // accessible after this wrapFuncToExecuteInWorkQueueBlockingImpl(...) function returns.
+
+        // Note also that we assume that the user's function accepts arguments by const reference.
+        // This will avoid an unnecessary copy when the surrounding code (e.g., the RPC server)
+        // calls the function returned here. Deeper inside our implementation, we will eventually
+        // copy these arguments so they are guaranteed to be accessible when we eventually execute
+        // the user's function, but we want to avoid copying wherever possible.
+        return [&work_queue, func](TArgs&... args) -> TReturn {
+
+            // Note that we capture func and args by value because we want to guarantee that they are both
+            // still accessible after scheduleAndExecuteTaskBlocking(...) returns. Strictly speaking, this
+            // guarantee is not necessary for this blocking implementation, but we capture by value anyway
+            // for consistency with our non-blocking implementation.
+            auto task = CopyConstructiblePackagedTask(
+                [func, args...]() -> TReturn {
+                    return func(args...);
+                });
+
+            std::future<TReturn> future = task.get_future();
+            boost::asio::post(work_queue.io_context_, std::move(task));
+            return future.get();
+            };
+    }
+
+    // The purpose of this class is to provide a copy-constructible type that derives from std::packaged_task. This is necessary 
+    // because boost::asio requires that task objects are copy-constructible, but std::packaged_task is not copy-constructible.
+    template <typename TFunc>
```

Comment:
I like being able to use `std::invoke_result_t<TFunc>` in situations where we need `TReturn`, but I liked it better when `CopyConstructiblePackagedTask` was templated on `TReturn` rather than `TFunc` because then it is more obvious that it is a drop-in replacement for `std::packaged_task`.

Let's keep using `std::invoke_result_t<TFunc>` because it's so clean and convenient, but let's use it in our `using` statements when we need to define `TReturn`.